File: packages/engine-elixir/lib/carFinder.ex
defmodule CarFinder do
  def rad(x) do
    x * Math.pi() / 180
  end

  @doc """

  Calculate birds distance (haversine) from a cars current position to a booking
  """
  def distance(car, %{departure: departure}) do
    distance = Distance.haversine(Car.position(car), departure)
    %{car: car, distance: distance}
  end

  @doc """
  Sort all cars in a list according to birds distance to a booking
  """
  def closestCars(booking, cars) do
    cars
    |> Enum.map(fn car -> distance(car, booking) end)
    |> Enum.sort(fn a, b -> a.distance < b.distance end)
    |> Enum.map(fn car -> car.car end)
  end

  @doc """
  Sort all cars in a list based on the smallest calculated detour from the cars original route
  """
  def detourCars(cars, booking) do
    cars
    |> Enum.map(fn car ->
      Osrm.trip([
        car.position,
        booking.departure,
        booking.destination,
        car.heading || car.position
      ])
      |> (fn %{code: "Ok", trips: [detour | _rest]} ->
            %{
              car: car,
              detour: detour
            }
          end).()
    end)
    |> Enum.map(fn %{car: car, detour: detour} ->
      %{
        car: car,
        booking: booking,
        # detour: Map.put(detour, :diff, Score.calculate(booking, car, detour))
        score: Score.calculate(booking, car, detour)
      }
    end)
    # |> Enum.sort(fn a, b -> a.detour.diff < b.detour.diff end)
    |> Enum.sort_by(fn a -> a.score end, :desc)
    |> Enum.sort_by(fn a -> a.car.busy end, :asc)

    # |> Enum.sort(fn a, b -> a.score < b.score end)
  end

  @doc """
  find the best suitable cars for a booking
  """
  def find(booking, cars) do
    closestCars(booking, cars)
    # |> detourCars(booking)
    |> detourCarsNew(booking)
    |> Enum.take(2)
  end

  def detourCarsNew(cars, booking) do
    cars
    |> Enum.map(fn car ->
      Car.calculateDetours(car, booking)
      |> (fn [first | _rest] -> %{car: car, booking: booking, score: first.score} end).()
    end)
    |> Enum.sort_by(fn a -> a.score end, :desc)
  end
end

File: packages/engine-js/lib/carFinder.js
const distance = require('./distance')
const { route, trip } = require('./osrm')
const _ = require('highland')

function estimateTimeToArrival(car, destination) {
  return route(car.position, destination)
    .then(route => {
      return {
        distance: route.distance,
        route: route,
        tta: route.duration || 9999,
        car: car,
      }
    })
    .catch(err => console.error('estimated route', err))
}

const closestCars = (booking, within = 5000, count = 50) =>
  _.pipeline(cars =>
    cars
      .map(car => ({
        distance: distance.haversine(car.position, booking.departure),
        car: car,
      }))
      .take(count)
      // .filter(hit => hit.distance < within)
      .sortBy((a, b) => a.distance - b.distance)
      .map(hit => hit.car)
      // .filter(car => !car.busy)
      .errors(err => console.error('closestCars', err))
  )

const fastestCars = booking =>
  _.pipeline(cars =>
    cars
      .ratelimit(5, 500)
      .flatMap(car => _(estimateTimeToArrival(car, booking.departure)))
      .errors(err => console.error('estimate err', err))
      .sortBy((a, b) => a.tta - b.tta)
      // .filter(hit => hit.tta < 15 * 60)
      .errors(err => console.error('fastestCars', err))
  )

const detourCars = booking =>
  _.pipeline(cars =>
    cars.flatMap(car =>
      _(
        trip([
          car.position,
          booking.departure,
          booking.destination,
          car.heading,
        ])
          .then(detour => ({
            car,
            detour: detour.code === 'Ok' ? detour.trips.shift() : null,
          }))
          .then(({ car, detour }) => ({
            car,
            detour: {
              ...detour,
              diff: detour.distance - car.heading.route.distance,
            },
          }))
      )
    )
  )

const findCars = (booking, cars) => {
  const carsCache = new Map()
  cars.fork().each(car => {
    carsCache.set(car.id, car)
  })

  return (
    _.merge([_(carsCache.values()), cars.fork()])
      // _.merge(_(carsCache.values()), _(carPositions).fork())
      .pipe(closestCars(booking))
      .pipe(detourCars(booking))
      /*.take(50)
    .tap(car => console.log('closest', car.id, car.tta))
    .pipe(fastestCars(booking))
    .tap(car => console.log('fastest', car.id, car.tta))*/
      .take(2)
      .sortBy((a, b) => a.detour.diff - b.detour.diff)
      .errors(err => console.error('findCars', err))
  )
}

module.exports = findCars
